docs/reusing-typechecking-results.md

```diff
@@ -0,0 +1,171 @@
+|Import non-system references        |  0,5475|  0,0775|    128|      0|      0|      0|    374|     30|
+|Typecheck                           |  3,1157|  2,5612|    198|     18|     15|      3|    712|     45|     <-- long
+|Typechecked                         |  3,1219|  0,0002|    198|      0|      0|      0|    712|     45|
+// main2
+|Write Interface File                |  3,1280|  0,0000|    198|      0|      0|      0|    712|     45|
+|Write XML doc signatures            |  3,1363|  0,0000|    198|      0|      0|      0|    712|     45|
+|Write XML docs                      |  3,1435|  0,0002|    198|      0|      0|      0|    712|     45|
+// main3
+|Encode Interface Data               |  3,1803|  0,0296|    198|      0|      0|      0|    712|     45|
+|Optimizations                       |  8,9949|  5,8049|    216|     43|     42|      2|    457|     45|     <-- longest
+|Ending Optimizations                |  9,0015|  0,0000|    216|      0|      0|      0|    457|     45|
+|Encoding OptData                    |  9,0090|  0,0010|    216|      0|      0|      0|    457|     45|
+|TailCall Checks                     |  9,0190|  0,0013|    216|      0|      0|      0|    457|     45|
+// main4, main5
+|TAST -> IL                          |  9,0463|  0,0210|    217|      0|      0|      0|    458|     46|
+// main6
+|Write .NET Binary                   |  9,2463|  0,1930|    218|      0|      0|      0|    464|     46|     <-- long
+--------------------------------------------------------------------------------------------------------
+```
+
+The phases taking the longest are marked in the right. Those are assembly import, type check, optimizations and IL writing.
+
+Optimizations are not relevant in the dev loop with run/debug/test cycles so we won't take those into account.
+
+## Implementation plan
+
+The conclusion from the above is that there is a lot of potential in caching - at the same time, implementing things to the full potential right away will require serious changes in the compiler code, which will be dangerous, hard to test and review. Therefore, I propose to implement the feature in stages where each stage brings gains in some scenarios.
+
+**Stage 1 - force-gen and compare typechecking graphs**
+
+We already create the TC graph in `main1` and we are able to dump it (in Mermaid) via the compiler flags.
+
+That means we can force-gen and save the graph which will allow us to skip retypechecking if we detect that the graph is not changed. We should also track all the compilation information (the argument string).
+
+This step won't bring big observable benefits, yet it will create necessary MSBuild hooks to communicate the intermediate files folder towards the compiler, add time-based and hash-based cache invalidation logic, and create testing rails which will include the `clean` and `rebuild` tests to make sure the cache is easily invalidated on demand.
```

Comment:
> time-based and hash-based cache invalidation logic

This is the fun part. The hash of each file should contain also all its dependencies. It's kinda the same problem [`FSharpProjectSnapshot`](https://github.com/dotnet/fsharp/blob/main/src/Compiler/Service/FSharpProjectSnapshot.fs) is trying to solve. It would be nice if we could somehow unify this.